packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnGrouping.ts

```diff
@@ -0,0 +1,177 @@
+            `- ${rep[key].join(' > ')}`,
+            `- ${value.join(' > ')}`,
+          ].join('\n'),
+        );
+      }
+      rep[key] = value;
+    });
+  });
+  return rep;
+};
+
+export const unwrapGroupingColumnModel = (
+  columnGroupingModel?: GridColumnGroupingModel,
+): UnwrappedGroupingModel => {
+  if (!columnGroupingModel) {
+    return {};
+  }
+
+  const rep: UnwrappedGroupingModel = {};
+
+  columnGroupingModel.forEach((columnGroupNode) => {
+    const unwrappedSubTree = recurrentUnwrapGroupingColumnModel(columnGroupNode, []);
+    Object.entries(unwrappedSubTree).forEach(([key, value]) => {
+      if (rep[key] !== undefined) {
+        throw new Error(
+          [
+            `MUI: column grouping has duplicated field.`,
+            `column field ${key} occurres two times in the grouping model:`,
+            `- ${rep[key].join(' > ')}`,
+            `- ${value.join(' > ')}`,
+          ].join('\n'),
+        );
+      }
+      rep[key] = value;
+    });
+  });
+
+  return rep;
+};
+
+const createGroupLookup = (columnGroupingModel: GridColumnNode[]): GridColumnGroupLookup => {
+  let groupLookup: GridColumnGroupLookup = {};
+
+  columnGroupingModel.forEach((node) => {
+    if (isLeaf(node)) {
+      return;
+    }
+    const { groupId, children, ...other } = node;
+    if (!groupId) {
+      throw new Error(
+        'MUI-DataGrid: an element of the columnGroupingModel does not have either `field` or `groupId`',
+      );
+    }
+    if (!children) {
+      console.warn(`MUI-DataGrid: group groupId=${groupId} has no children. `);
+    }
+    const groupParam = { ...other, groupId };
+    const subTreeLookup = createGroupLookup(children);
+    if (subTreeLookup[groupId] !== undefined || groupLookup[groupId] !== undefined) {
+      throw new Error(`MUI: The groupId ${groupId} is used multiple times`);
+    }
+    groupLookup = { ...groupLookup, ...subTreeLookup, [groupId]: groupParam };
+  });
+
+  return { ...groupLookup };
+};
+export const columnGroupsStateInitializer: GridStateInitializer<
+  Pick<DataGridProcessedProps, 'columnGroupingModel'>
+> = (state, props) => {
+  const groupLookup = createGroupLookup(props.columnGroupingModel ?? []);
+  return {
+    ...state,
+    columnGrouping: { lookup: groupLookup, groupCollapsedModel: {} },
+  };
+};
+
+/**
+ * @requires useGridColumns (method, event)
+ * @requires useGridParamsApi (method)
+ */
+export const useGridColumnGrouping = (
+  apiRef: React.MutableRefObject<GridApiCommunity>,
+  props: Pick<DataGridProcessedProps, 'columnGroupingModel'>,
+) => {
+  /**
+   * API METHODS
+   */
+  const getColumnGroupPath = React.useCallback<GridColumnGroupingApi['getColumnGroupPath']>(
+    (field) => {
+      const columnLookup = gridColumnLookupSelector(apiRef);
+
+      return columnLookup[field]?.groupPath ?? [];
+    },
+    [apiRef],
+  );
+
+  const getAllGroupDetails = React.useCallback<GridColumnGroupingApi['getAllGroupDetails']>(() => {
```

Comment:
Maybe wait before adding a method that we never need to use internally (can be replaced by the selector)
Espacially if we are going to open the `apiRef` to the community plan

Same for the other two methods
We don't have a `getColumnComputedWidth` for example.
Let's keep the api surface as small as possible unless people need it.

Reply:
made them `unstable_`